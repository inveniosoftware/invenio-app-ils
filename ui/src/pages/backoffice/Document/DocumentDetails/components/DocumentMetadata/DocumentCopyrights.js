import { MetadataTable } from '@pages/backoffice';
import capitalize from 'lodash/capitalize';
import React, { Component } from 'react';
import PropTypes from 'prop-types';

export class DocumentCopyrights extends Component {
  renderCopyrights() {
    const { copyrights } = this.props.document.metadata;
    const rows = [];
    if (copyrights) {

      for (const [key, val] of Object.entries(copyrights)) {
        rows.push({name: capitalize(key), value: val});
      }
    }
    return rows;
  }

  render() {
      return <MetadataTable rows={this.renderCopyrights()} />;
  }
}

DocumentCopyrights.propTypes = {
  document: PropTypes.object.isRequired,
};
