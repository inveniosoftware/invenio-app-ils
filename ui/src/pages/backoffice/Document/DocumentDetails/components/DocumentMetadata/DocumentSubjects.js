import { MetadataTable } from '@pages/backoffice';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { List } from 'semantic-ui-react';
import { groupBy, isEmpty } from 'lodash';

export class DocumentSubjects extends Component {
  render() {
    const { document } = this.props;

    if (!isEmpty(document.metadata.subjects)) {
      const groupedSubjects = groupBy(document.metadata.subjects, 'scheme');
      let rows = [];
      for (const [scheme, idsList] of Object.entries(groupedSubjects)) {
        rows.push({
          name: scheme,
          value: (
            <List bulleted>
              {idsList.map((entry, idx) => (
                <List.Item key={idx}>
                  <List.Content>{entry.value}</List.Content>
                </List.Item>
              ))}
            </List>
          ),
        });
      }
      return <MetadataTable rows={rows} />;
    }
    return null;
  }
}

DocumentSubjects.propTypes = {
  document: PropTypes.object.isRequired,
};
