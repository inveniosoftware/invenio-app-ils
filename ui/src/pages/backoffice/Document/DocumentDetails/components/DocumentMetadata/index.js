import { connect } from 'react-redux';

import DocumentMetadataComponent from './DocumentMetadata';
import {
  deleteDocument,
  fetchDocumentDetails,
  updateDocument,
  requestLoanForPatron,
} from '../../state/actions';

const mapStateToProps = state => ({
  documentDetails: state.documentDetails.data,
  error: state.documentDetails.error,
  relations: state.documentRelations.data,
});

const mapDispatchToProps = dispatch => ({
  fetchDocumentDetails: documentPid =>
    dispatch(fetchDocumentDetails(documentPid)),
  deleteDocument: documentPid => dispatch(deleteDocument(documentPid)),
  updateDocument: (documentPid, path, value) =>
    dispatch(updateDocument(documentPid, path, value)),
  requestLoanForPatron: (documentPid, patronPid, optionalParams = {}) =>
    dispatch(requestLoanForPatron(documentPid, patronPid, optionalParams)),
});

export const DocumentMetadata = connect(
  mapStateToProps,
  mapDispatchToProps
)(DocumentMetadataComponent);
