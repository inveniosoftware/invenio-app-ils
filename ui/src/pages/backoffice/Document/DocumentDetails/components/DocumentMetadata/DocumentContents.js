import { DocumentAbstract, DocumentToc } from '@components/Document';
import { DocumentSubjects } from './DocumentSubjects';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Divider, Header } from 'semantic-ui-react';
import isEmpty from 'lodash/isEmpty';

export class DocumentContents extends Component {
  render() {
    const { document } = this.props;
    return (
      <>
        <Header as="h3">Abstract</Header>
        <DocumentAbstract metadata={document.metadata} lines={10} />

        {!isEmpty(document.metadata.table_of_content) && (
          <>
            <Divider />
            <Header as="h3">Table of content</Header>
            <DocumentToc document={document} />
          </>
        )}

        {!isEmpty(document.metadata.subjects) && (
          <>
            <Divider />
            <Header as="h3">Subjects</Header>
            <DocumentSubjects document={document} />
          </>
        )}
      </>
    );
  }
}

DocumentContents.propTypes = {
  document: PropTypes.object.isRequired,
};
