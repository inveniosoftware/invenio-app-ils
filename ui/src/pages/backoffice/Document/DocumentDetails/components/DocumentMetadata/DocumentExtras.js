import { MetadataTable } from '@pages/backoffice/components';
import capitalize from 'lodash/capitalize';
import isEmpty from 'lodash/isEmpty';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Header, List } from 'semantic-ui-react';

export class DocumentExtras extends Component {
  renderConferenceInfo = () => {
    const { conference_info } = this.props.document.metadata;
    let rows = [];
    for (const [key, val] of Object.entries(conference_info)) {
      if (Array.isArray(val)) {
        const arrayVals = (
          <List>
            {val.map((entry, idx) => (
              <List.Item key={idx}>
                <List.Content>
                  <List.Header>{entry.scheme}</List.Header>
                  <List.Description>{entry.value}</List.Description>
                </List.Content>
              </List.Item>
            ))}
          </List>
        );
        rows.push({ name: capitalize(key), value: arrayVals });
      } else {
        rows.push({ name: capitalize(key), value: val });
      }
    }
    return rows;
  };

  render() {
    const { document } = this.props;
    return (
      <>
        {!isEmpty(document.metadata.conference_info) && (
          <>
            <Header>Conference info</Header>
            <MetadataTable rows={this.renderConferenceInfo()} />
          </>
        )}
      </>
    );
  }
}

DocumentExtras.propTypes = {
  document: PropTypes.object.isRequired,
};
