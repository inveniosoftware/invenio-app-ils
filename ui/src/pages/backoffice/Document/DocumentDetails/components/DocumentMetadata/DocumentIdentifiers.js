import { InfoMessage, MetadataTable } from '@pages/backoffice/components';
import { groupedSchemeValueList } from '@pages/backoffice/utils';
import PropTypes from 'prop-types';
import React, { Component } from 'react';

export class DocumentIdentifiers extends Component {
  render() {
    const { document } = this.props;
    return document.metadata.identifiers ? (
      <MetadataTable
        rows={groupedSchemeValueList(document.metadata.identifiers)}
      />
    ) : (
      <InfoMessage
        header={'No stored identifiers.'}
        content={'Edit document to add identifiers'}
      />
    );
  }
}

DocumentIdentifiers.propTypes = {
  document: PropTypes.object.isRequired,
};
