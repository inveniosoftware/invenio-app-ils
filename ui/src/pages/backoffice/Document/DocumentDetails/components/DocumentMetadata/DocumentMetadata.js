import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import {
  Grid,
  Segment,
  Container,
  Header,
  List,
  Icon,
  Button,
  Popup,
} from 'semantic-ui-react';
import PropTypes from 'prop-types';
import { invenioConfig } from '@config';
import { MetadataTable } from '@pages/backoffice/components/MetadataTable';
import { EditButton, NewButton } from '@pages/backoffice/components/buttons';
import {
  document as documentApi,
  documentRequest as documentRequestApi,
  loan as loanApi,
  item as itemApi,
  patron as patronApi,
} from '@api';
import { BackOfficeRoutes } from '@routes/urls';
import { DeleteRecordModal } from '@pages/backoffice/components/DeleteRecordModal';
import { ESSelectorLoanRequest } from '@components/ESSelector';
import { serializePatron } from '@components/ESSelector/serializer';
import { formatPidTypeToName } from '@pages/backoffice/components/ManageRelationsButton/utils';
import { isEmpty } from 'lodash';

export default class DocumentMetadata extends Component {
  constructor(props) {
    super(props);
    this.deleteDocument = props.deleteDocument;
    this.documentPid = props.documentDetails.metadata.pid;
  }

  requestLoan = (
    patronPid,
    { requestEndDate = null, deliveryMethod = null } = {}
  ) => {
    const documentPid = this.props.documentDetails.metadata.pid;
    const optionalParams = {};
    if (!isEmpty(requestEndDate)) {
      optionalParams.requestEndDate = requestEndDate;
    }
    if (!isEmpty(deliveryMethod)) {
      optionalParams.deliveryMethod = deliveryMethod;
    }
    this.props.requestLoanForPatron(documentPid, patronPid, optionalParams);
  };

  renderTags(tags) {
    if (!isEmpty(tags)) {
      return (
        <List horizontal>
          {tags.map(tag => (
            <List.Item key={tag}>
              <Link
                to={BackOfficeRoutes.documentsListWithQuery(
                  documentApi
                    .query()
                    .withTag(tag)
                    .qs()
                )}
              >
                {tag.name}
              </Link>
            </List.Item>
          ))}
        </List>
      );
    } else {
      return null;
    }
  }

  handleOnLoanRefClick(loanPid) {
    const navUrl = BackOfficeRoutes.loanDetailsFor(loanPid);
    window.open(navUrl, `_loan_${loanPid}`);
  }

  handleOnItemRefClick(itemPid) {
    const navUrl = BackOfficeRoutes.itemDetailsFor(itemPid);
    window.open(navUrl, `_item_${itemPid}`);
  }

  handleOnRequestRefClick(docReqPid) {
    const navUrl = BackOfficeRoutes.documentRequestDetailsFor(docReqPid);
    window.open(navUrl, `_document_request_${docReqPid}`);
  }

  async getRelationRefs() {
    const hits = [];
    for (const [relation, records] of Object.entries(this.props.relations)) {
      for (const record of records) {
        const type = formatPidTypeToName(record.pid_type);
        hits.push({
          id: `${type} ${record.pid} (${relation})`,
        });
      }
    }
    const obj = {
      data: {
        hits: hits,
        total: hits.length,
      },
    };
    return obj;
  }

  createRefProps(documentPid) {
    const loanStates = invenioConfig.circulation.loanRequestStates.concat(
      invenioConfig.circulation.loanActiveStates
    );

    const loanRefProps = {
      refType: 'Loan',
      onRefClick: this.handleOnLoanRefClick,
      getRefData: () =>
        loanApi.list(
          loanApi
            .query()
            .withDocPid(documentPid)
            .withState(loanStates)
            .qs()
        ),
    };

    const itemRefProps = {
      refType: 'Items',
      onRefClick: this.handleOnItemRefClick,
      getRefData: () =>
        itemApi.list(
          itemApi
            .query()
            .withDocPid(documentPid)
            .qs()
        ),
    };

    const relationRefProps = {
      refType: 'Related',
      onRefClick: () => {},
      getRefData: () => this.getRelationRefs(),
    };

    const requestRefProps = {
      refType: 'DocumentRequest',
      onRefClick: this.handleOnRequestRefClick,
      getRefData: () =>
        documentRequestApi.list(
          documentRequestApi
            .query()
            .withDocPid(documentPid)
            .qs()
        ),
    };

    return [loanRefProps, itemRefProps, relationRefProps, requestRefProps];
  }

  renderHeader(document) {
    return (
      <Grid.Row>
        <Grid.Column width={9} verticalAlign={'middle'}>
          <Header as="h1">
            Document #{document.pid} - {document.metadata.title}
          </Header>
        </Grid.Column>
        <Grid.Column width={7} textAlign="right">
          <NewButton
            text="New item"
            to={{
              pathname: BackOfficeRoutes.itemCreate,
              state: { document },
            }}
          />
          <NewButton
            text="New e-item"
            to={{
              pathname: BackOfficeRoutes.eitemCreate,
              state: { document },
            }}
          />
          <EditButton to={BackOfficeRoutes.documentEditFor(this.documentPid)} />
          <DeleteRecordModal
            deleteHeader={`Are you sure you want to delete the Document
            record with ID ${document.pid}?`}
            refProps={this.createRefProps(document.pid)}
            onDelete={() => this.deleteDocument(document.pid)}
          />
        </Grid.Column>
      </Grid.Row>
    );
  }

  prepareData(document) {
    const rows = [
      { name: 'Title', value: document.metadata.title },
      {
        name: 'Authors',
        value: document.metadata.authors
          ? document.metadata.authors.map(author => author.full_name).join(',')
          : '',
      },
      {
        name: 'Keywords',
        value: document.metadata.keywords
          ? document.metadata.keywords.value +
            ' (' +
            document.metadata.keywords.source +
            ')'
          : null,
      },
      {
        name: 'Tags',
        value: this.renderTags(document.metadata.tags),
      },
    ];
    const request = document.metadata.request;
    if (!isEmpty(request)) {
      rows.push({
        name: 'Document Request',
        value: (
          <Link to={BackOfficeRoutes.documentRequestDetailsFor(request.pid)}>
            {request.state}
          </Link>
        ),
      });
    }
    return rows;
  }

  requestLoanButton = (
    <div>
      <Button
        positive
        icon="add"
        labelPosition="left"
        size="small"
        content="New Loan Request"
      />
      <Popup
        content="Request a loan on this document on behalf of a patron"
        trigger={<Icon name="info circle" size="large" />}
      />
    </div>
  );

  onSelectPatronResult = result => {
    const email = result.id.toLowerCase();
    result.metadata = { email: email };
    result.id = email;
    result.key = email;
    result.title = email;
  };

  render() {
    const document = this.props.documentDetails;
    const rows = this.prepareData(document);
    return (
      <Segment className="document-metadata">
        <Grid padded columns={2}>
          {this.renderHeader(document)}
          <ESSelectorLoanRequest
            trigger={this.requestLoanButton}
            query={patronApi.list}
            serializer={serializePatron}
            title={`Request a loan for document ${document.pid}`}
            content={
              'Search for the patron to whom the loan should be assigned'
            }
            selectionInfoText={
              'The loan will be assigned to the following patron'
            }
            emptySelectionInfoText={'No patrons selected yet'}
            onSave={this.requestLoan}
            saveButtonContent={'Perform request'}
          />
          <Grid.Row>
            <Grid.Column>
              <MetadataTable rows={rows} />
            </Grid.Column>
            <Grid.Column>
              <Container>
                <Header as="h3">Abstract</Header>
                <p>{document.metadata.abstract}</p>
              </Container>
            </Grid.Column>
          </Grid.Row>
        </Grid>
      </Segment>
    );
  }
}

DocumentMetadata.propTypes = {
  documentDetails: PropTypes.object.isRequired,
};
