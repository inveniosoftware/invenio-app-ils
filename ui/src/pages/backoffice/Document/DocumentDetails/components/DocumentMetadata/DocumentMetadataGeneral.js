import { toShortDate } from '@api/date';
import { DocumentAuthors } from '@components';
import { DocumentLanguages, DocumentTags } from '@components/Document';
import { MetadataTable } from '@pages/backoffice';
import { UrlList } from '@pages/backoffice/components/UrlList';
import { BackOfficeRoutes } from '@routes/urls';
import { isEmpty } from 'lodash';
import get from 'lodash/get';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { Container, Divider, Header } from 'semantic-ui-react';

export class DocumentMetadataGeneral extends Component {
  prepareGeneral = () => {
    const { document } = this.props;
    const urls = get(this.props, 'document.metadata.urls', []);

    const rows = [
      { name: 'Title', value: document.metadata.title },
      {
        name: 'Authors',
        value: <DocumentAuthors metadata={document.metadata} />,
      },
      { name: 'Publication year', value: document.metadata.publication_year },
      {
        name: 'Keywords',
        value: document.metadata.keywords
          ? `${document.metadata.keywords.value} (${document.metadata.keywords.source})`
          : null,
      },
      {
        name: 'Tags',
        value: <DocumentTags size="mini" metadata={document.metadata} />,
      },
      {
        name: 'Edition',
        value: document.metadata.edition,
      },
      {
        name: 'Languages',
        value: <DocumentLanguages metadata={document.metadata} />,
      },
      {
        name: 'Urls',
        value: <UrlList urls={urls} />,
      },
    ];

    const request = document.metadata.request;
    if (!isEmpty(request)) {
      rows.push({
        name: 'Document Request',
        value: (
          <Link to={BackOfficeRoutes.documentRequestDetailsFor(request.pid)}>
            {request.state}
          </Link>
        ),
      });
    }
    return rows;
  };

  prepareImprintInfo = () => {
    const { document } = this.props;

    return [
      { name: 'Publisher', value: document.metadata.imprint.publisher },
      { name: 'Date', value: toShortDate(document.metadata.imprint.date) },
      { name: 'Place', value: document.metadata.imprint.place },
      { name: 'Reprint date', value: document.metadata.imprint.reprint_date },
    ];
  };

  render() {
    const { document } = this.props;
    return (
      <Container fluid>
        <MetadataTable rows={this.prepareGeneral()} />

        {!isEmpty(document.metadata.imprint) && (
          <>
            <Divider />
            <Header as="h3">Imprint</Header>
            <MetadataTable rows={this.prepareImprintInfo()} />
          </>
        )}
      </Container>
    );
  }
}

DocumentMetadataGeneral.propTypes = {
  document: PropTypes.object.isRequired,
};
