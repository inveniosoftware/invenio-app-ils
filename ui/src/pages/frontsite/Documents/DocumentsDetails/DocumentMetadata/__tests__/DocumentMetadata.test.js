import React from 'react';
import { mount } from 'enzyme';
import * as testData from '@testData/documents.json';
import DocumentMetadata from '../DocumentMetadata';

jest.mock('../../../../components/Document', () => {
  return {
    DocumentMetadataAccordion: () => null,
    DocumentMetadataTabs: () => null,
    DocumentStats: () => null,
  };
});

describe('DocumentMetadata tests', () => {
  let component;

  afterEach(() => {
    if (component) {
      component.unmount();
    }
  });

  const documentDetails = {
    id: 71,
    created: '2019-07-08T10:44:02.366+02:00',
    updated: '2019-07-08T10:44:18.354+02:00',
    links: { self: 'https://127.0.0.1:5000/api/documents/71' },
    metadata: {
      $schema: 'https://127.0.0.1:5000/schemas/documents/document-v1.0.0.json',
      _access: {},
      ...testData[0],
      relations: {},
      circulation: {},
      pid: '71',
      document_type: ['BOOK'],
      tag_pids: ['24'],
      tags: [],
      eitems: {
        total: 1,
        hits: [
          {
            description: 'Non ipsum',
            pid: '145',
            open_access: true,
          },
        ],
      },
    },
  };

  it('should render the document correctly', () => {
    component = mount(<DocumentMetadata documentDetails={documentDetails} />);
    expect(component).toMatchSnapshot();

    const rows = component
      .find('DocumentMetadata')
      .find('Container')
      .filterWhere(
        element => element.prop('data-test') === documentDetails.metadata.pid
      );
    expect(rows).toHaveLength(1);
  });
});
