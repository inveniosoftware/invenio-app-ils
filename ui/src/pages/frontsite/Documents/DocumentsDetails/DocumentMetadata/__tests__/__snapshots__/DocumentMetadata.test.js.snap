// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentMetadata tests should render the document correctly 1`] = `
<DocumentMetadata
  documentDetails={
    Object {
      "created": "2019-07-08T10:44:02.366+02:00",
      "id": 71,
      "links": Object {
        "self": "https://127.0.0.1:5000/api/documents/71",
      },
      "metadata": Object {
        "$schema": "https://127.0.0.1:5000/schemas/documents/document-v1.0.0.json",
        "_access": Object {},
        "abstract": "This is an abstract",
        "authors": Array [
          Object {
            "full_name": "Jack E. Davis",
          },
        ],
        "circulation": Object {},
        "document_type": Array [
          "BOOK",
        ],
        "edition": "1",
        "eitems": Object {
          "hits": Array [
            Object {
              "description": "Non ipsum",
              "open_access": true,
              "pid": "145",
            },
          ],
          "total": 1,
        },
        "keywords": Object {
          "source": "X",
          "value": "Patata",
        },
        "languages": Array [
          "en",
        ],
        "pid": "71",
        "relations": Object {},
        "tags": Array [],
        "title": "The Gulf: The Making of An American Sea",
      },
      "updated": "2019-07-08T10:44:18.354+02:00",
    }
  }
>
  <Container
    className="document-metadata"
    data-test="71"
  >
    <div
      className="ui container document-metadata"
      data-test="71"
    >
      <Responsive
        getWidth={[Function]}
        minWidth={768}
      >
        <div>
          <DocumentMetadataTabs />
        </div>
      </Responsive>
      <Responsive
        getWidth={[Function]}
        maxWidth={767}
        minWidth={320}
      />
      <Container
        textAlign="center"
      >
        <div
          className="ui center aligned container"
        >
          <DocumentStats
            document={
              Object {
                "created": "2019-07-08T10:44:02.366+02:00",
                "id": 71,
                "links": Object {
                  "self": "https://127.0.0.1:5000/api/documents/71",
                },
                "metadata": Object {
                  "$schema": "https://127.0.0.1:5000/schemas/documents/document-v1.0.0.json",
                  "_access": Object {},
                  "abstract": "This is an abstract",
                  "authors": Array [
                    Object {
                      "full_name": "Jack E. Davis",
                    },
                  ],
                  "circulation": Object {},
                  "document_type": Array [
                    "BOOK",
                  ],
                  "edition": "1",
                  "eitems": Object {
                    "hits": Array [
                      Object {
                        "description": "Non ipsum",
                        "open_access": true,
                        "pid": "145",
                      },
                    ],
                    "total": 1,
                  },
                  "keywords": Object {
                    "source": "X",
                    "value": "Patata",
                  },
                  "languages": Array [
                    "en",
                  ],
                  "pid": "71",
                  "relations": Object {},
                  "tags": Array [],
                  "title": "The Gulf: The Making of An American Sea",
                },
                "updated": "2019-07-08T10:44:18.354+02:00",
              }
            }
          />
        </div>
      </Container>
    </div>
  </Container>
</DocumentMetadata>
`;
